python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py

```diff
@@ -2006,3 +2006,43 @@ def _specified_elsewhere() -> int:
             return my_op()
 
         assert _specified_elsewhere  # appease linter
+
+
+def test_multi_asset_specs_no_loop_construction() -> None:
+    upstream_asset = AssetSpec("upstream_asset")
+    downstream_asset = AssetSpec("downstream_asset", deps=[upstream_asset])
+
+    @multi_asset(specs=[upstream_asset])
+    def _upstream_op() -> Any:
+        raise NotImplementedError()
+
+    @multi_asset(specs=[downstream_asset])
+    def _downstream_op() -> Any:
+        raise NotImplementedError()
+
+    defs = Definitions(assets=[_upstream_op, _downstream_op])
+
+    assert defs.get_implicit_global_asset_job_def().asset_layer.asset_deps[
+        AssetKey("downstream_asset")
+    ] == {AssetKey("upstream_asset")}
+
+
+def test_multi_asset_specs_factory_esque_code() -> None:
+    upstream_asset = AssetSpec("upstream_asset")
+    downstream_asset = AssetSpec("downstream_asset", deps=[upstream_asset])
+
+    assets_defs: List[AssetsDefinition] = []
+
+    for i, spec in enumerate([upstream_asset, downstream_asset]):
+
+        @multi_asset(specs=[spec], name=f"_an_asset_op_{i}")
```

Comment:
The issue in this test was because the multi_assets were given the same name. The assets are auto-assigned aliases, but the underlying nodes still have the same name. we also maintain a mapping of node name to node definition, but since both assets have the same underlying node name, only one NodeDefinition is kept around (and it's downstream asset, i think because it was created last). 

This causes collisions when we validate dependencies because when we validate the deps for `_an_asset_op` (made from the `downstream_asset` spec) we want to get the outputs for `_an_asset_op` (made from the `upstream_asset` spec) the mapping of node name to NodeDefinition gives us the node for `_an_asset_op` (made from the `downstream_asset` spec), which doesnt have an `upstream_asset` output

you get exactly the same issue if you do this:
```python
@multi_asset(
        outs={"x": AssetOut(), "y": AssetOut()}
    )
    def a():
        return 1, 2

    @multi_asset(
        outs={"i": AssetOut(), "j": AssetOut()}
    )
    def a(): # redefined a!
        return 3, 4

    @asset(
        ins={"x": AssetIn(), "y": AssetIn()}
    )
    def add(x, y):
        return x + y

    defs = Definitions(assets=[a, add])
```

I don't think we can solve this with a backend change because of the op alias case. If we have this
```python
from dagster import op, job

    @op
    def generic_adder(x):
        return x + 1

    @op
    def one():
        return 1

    @op
    def two():
        return 2

    @job
    def do_math():
        generic_adder(one())
        generic_adder(two())
```

We want both `generic_adder`s to be assigned aliases, but still point to the same underlying node

Reply:
Ah you are right. Thanks for clarifying!